File: Core/Architect.API.Core.DataFactory/_Generated/Navigation.cs
using Architect.DataFactory;
using Architect.Utilities.Extensions;
using System;
using System.Collections.Generic;
using System.Data;
using DbType = Architect.DataFactory.Enumerations.DbType;

namespace Architect.API.Core.DataAccess.General
{
    /// <summary>
    /// Navegación disponible.
    /// </summary>
    public static partial class Navigation
    {

        /// <summary>
        /// Crea un registro en la tabla Navigation.
        /// </summary>
        /// <param name="navigationItem">Instancia de Navigation</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros creados.</returns>
        public static int Create(Architect.API.Core.Contracts.General.Navigation navigationItem, IDbConnection connection = null)
        {
            if (navigationItem.UpdateDate.IsEmpty())
            {
                navigationItem.UpdateDate = DateTime.Now;
            }
            return Database.Insert("INSERT INTO Navigation (NavigationId, CompanyId, ParentCode, Code, Type, Title, Description, MenuTitle, URLPath, URLHelp, SmallImage, BigImage, Sequence, RecordStatus, UpdateUserCode, UpdateDate) " +
                                                 "VALUES(:NavigationId, :CompanyId, :ParentCode, :Code, :Type, :Title, :Description, :MenuTitle, :URLPath, :URLHelp, :SmallImage, :BigImage, :Sequence, :RecordStatus, :UpdateUserCode, :UpdateDate)")
                            .AddParameter("NavigationId", DbType.Decimal, 9, navigationItem.NavigationId)
                            .AddParameter("CompanyId", DbType.Decimal, 5, navigationItem.CompanyId)
                            .AddParameter("ParentCode", DbType.AnsiStringFixedLength, 8, navigationItem.ParentCode)
                            .AddParameter("Code", DbType.AnsiStringFixedLength, 8, navigationItem.Code)
                            .AddParameter("Type", DbType.Decimal, 5, navigationItem.Type)
                            .AddParameter("Title", DbType.AnsiString, 120, navigationItem.Title)
                            .AddParameter("Description", DbType.AnsiString, 255, navigationItem.Description)
                            .AddParameter("MenuTitle", DbType.AnsiString, 40, navigationItem.MenuTitle)
                            .AddParameter("URLPath", DbType.AnsiString, 255, navigationItem.URLPath)
                            .AddParameter("URLHelp", DbType.AnsiString, 255, navigationItem.URLHelp)
                            .AddParameter("SmallImage", DbType.AnsiString, 80, navigationItem.SmallImage)
                            .AddParameter("BigImage", DbType.AnsiString, 80, navigationItem.BigImage)
                            .AddParameter("Sequence", DbType.Decimal, 5, navigationItem.Sequence)
                            .AddParameter("RecordStatus", DbType.Decimal, 5, navigationItem.RecordStatus)
                            .AddParameter("UpdateUserCode", DbType.Decimal, 9, navigationItem.UpdateUserCode)
                            .AddParameter("UpdateDate", DbType.DateTime, 0, navigationItem.UpdateDate)
                            .Execute(connection, "Research");
        }

        /// <summary>
        /// Crea una lista de registros en la tabla Navigation.
        /// </summary>
        /// <remarks>Complemento para procesamiento masivo</remarks>
        /// <param name="navigationItems">Lista de instancia de Navigation</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista con el resultado de la creación de cada instancia.</returns>
        public static List<int> Create(List<Architect.API.Core.Contracts.General.Navigation> navigationItems, IDbConnection connection = null)
        {
            List<int> result = new List<int>();
            bool local = false;

            if (connection == null)
            {
                connection = Architect.DataFactory.Database.OpenConnection("Research");
                local = true;
            }
            foreach (Architect.API.Core.Contracts.General.Navigation item in navigationItems)
            {
                result.Add(Create(item, connection));
            }
            if (local)
            {
                connection.Close();
            }
            return result;
        }

        /// <summary>
        /// Recupera un registro en la tabla Navigation por medio de su clave primaria.
        /// </summary>
        /// <param name="navigationid">Identificación única del registro.</param>
        /// <param name="companyId">Identificación de la compañia propietaria.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Instancia de Navigation</returns>
        public static Architect.API.Core.Contracts.General.Navigation Retrieve(int navigationid, int companyId, IDbConnection connection = null)
        {
            Architect.API.Core.Contracts.General.Navigation result = null;
            Database.Select("SELECT NavigationId, Navigation.CompanyId, ParentCode, Code, Type, Title, Description, MenuTitle, URLPath, URLHelp, SmallImage, BigImage, Sequence, Navigation.RecordStatus, Navigation.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Navigation.UpdateDate " +
                              "FROM Navigation LEFT JOIN UserMember um ON um.UserId = Navigation.UpdateUserCode " +
                             "WHERE Navigation.NavigationId=:NavigationId AND Navigation.CompanyId=:CompanyId")
                        .AddParameter("NavigationId", DbType.Decimal, 9, navigationid)
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {
                            result = DataReaderToNavigation(reader);
                        }));
            return result;
        }

        /// <summary>
        /// Recupera una lista de registros en la tabla Navigation.
        /// </summary>
        /// <param name="companyId">Identificación de la compañia propietaria.</param>
        /// <param name="filter">Filtro personalizado.</param>
        /// <param name="parameters">Lista de parametros para complemento.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista de instancias de Navigation</returns>
        public static List<Architect.API.Core.Contracts.General.Navigation> RetrieveAll(int companyId, string filter, List<DataFactory.Contracts.Parameter> parameters = null, IDbConnection connection = null)
        {
            List<Architect.API.Core.Contracts.General.Navigation> result = new List<Architect.API.Core.Contracts.General.Navigation>();
            Database.Select("SELECT NavigationId, Navigation.CompanyId, ParentCode, Code, Type, Title, Description, MenuTitle, URLPath, URLHelp, SmallImage, BigImage, Sequence, Navigation.RecordStatus, Navigation.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Navigation.UpdateDate " +
                              "FROM Navigation LEFT JOIN UserMember um ON um.UserId = Navigation.UpdateUserCode " +
                             "WHERE Navigation.CompanyId=:CompanyId" + filter)
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .AddParameter(parameters)
                        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {
                            result.Add(DataReaderToNavigation(reader));
                        }));
            return result;
        }

        /// <summary>
        /// Recupera una lista de registros en la tabla Navigation.
        /// </summary>
        /// <param name="companyId">Identificación de la compañia propietaria.</param>
        /// <param name="filter">Filtro personalizado.</param>
        /// <param name="beginIndex">Indice inicial para el paginado.</param>
        /// <param name="endIndex">Indice final para el paginado.</param>
        /// <param name="parameters">Lista de parametros para complemento.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista de instancias de Navigation</returns>
        public static List<Architect.API.Core.Contracts.General.Navigation> RetrieveAll(int companyId, string filter, int beginIndex, int endIndex, List<DataFactory.Contracts.Parameter> parameters = null, IDbConnection connection = null)
        {
            List<Architect.API.Core.Contracts.General.Navigation> result = new List<Architect.API.Core.Contracts.General.Navigation>();
            if (beginIndex == 0)
            {
                beginIndex = 1;
            }
            if (endIndex == 0 || endIndex < beginIndex)
            {
                endIndex = int.MaxValue;
            }
            Database.Select("SELECT * FROM (" +
                            "SELECT NavigationId, Navigation.CompanyId, ParentCode, Code, Type, Title, Description, MenuTitle, URLPath, URLHelp, SmallImage, BigImage, Sequence, Navigation.RecordStatus, Navigation.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Navigation.UpdateDate " +
                                   ", ROW_NUMBER() OVER (ORDER BY Navigation.Sequence ASC) RowNumber " +
                              "FROM Navigation LEFT JOIN UserMember um ON um.UserId = Navigation.UpdateUserCode " +
                             "WHERE Navigation.CompanyId=:CompanyId" + filter +
                               ") WHERE RowNumber BETWEEN :beginIndex AND :endIndex")
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .AddParameter(parameters)
                        .AddParameter("beginIndex", DbType.Decimal, 9, beginIndex)
                        .AddParameter("endIndex", DbType.Decimal, 9, endIndex)
                        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {
                            result.Add(DataReaderToNavigation(reader));
                        }));
            return result;
        }

        /// <summary>
        /// Genera complemento del query para habilitar el filtro por columnas establecidas.
        /// </summary>
        /// <param name="filter">Permite filtrar la lista de registros.</param>
        /// <param name="where">Indica que el complement debe incluir el WHERE del comando.</param>
        /// <returns>Complemento del query</returns>
        public static string FilterBuilder(string filter, bool includeWhere = true)
        {
            string result = string.Empty;


            return result;
        }

        /// <summary>
        /// Último valor asignado a clave unica de la tabla Navigation.
        /// </summary>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Último valor asignado.</returns>
        public static int RetrieveLastKey(IDbConnection connection = null)
        {

            return (int)Database.Select("SELECT NVL(MAX(NavigationId),0) " +
                                     "FROM Navigation")
                                .QueryScalar<Decimal>(connection, "Research");
        }

        /// <summary>
        /// Actualiza un registro en la tabla Navigation por medio de su clave primaria.
        /// </summary>
        /// <param name="navigationItem">Instancia de Navigation</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros actualizados.</returns>
        public static int Update(Architect.API.Core.Contracts.General.Navigation navigationItem, IDbConnection connection = null)
        {
            if (navigationItem.UpdateDate.IsEmpty())
            {
                navigationItem.UpdateDate = DateTime.Now;
            }
            return Database.Update("UPDATE Navigation " +
                                      "SET CompanyId=:CompanyId, ParentCode=:ParentCode, Code=:Code, Type=:Type, Title=:Title, Description=:Description, MenuTitle=:MenuTitle, URLPath=:URLPath, URLHelp=:URLHelp, SmallImage=:SmallImage, BigImage=:BigImage, Sequence=:Sequence, RecordStatus=:RecordStatus, UpdateUserCode=:UpdateUserCode, UpdateDate=:UpdateDate " +
                                    "WHERE NavigationId=:NavigationId")
                                .AddParameter("CompanyId", DbType.Decimal, 5, navigationItem.CompanyId)
                                .AddParameter("ParentCode", DbType.AnsiStringFixedLength, 8, navigationItem.ParentCode)
                                .AddParameter("Code", DbType.AnsiStringFixedLength, 8, navigationItem.Code)
                                .AddParameter("Type", DbType.Decimal, 5, navigationItem.Type)
                                .AddParameter("Title", DbType.AnsiString, 120, navigationItem.Title)
                                .AddParameter("Description", DbType.AnsiString, 255, navigationItem.Description)
                                .AddParameter("MenuTitle", DbType.AnsiString, 40, navigationItem.MenuTitle)
                                .AddParameter("URLPath", DbType.AnsiString, 255, navigationItem.URLPath)
                                .AddParameter("URLHelp", DbType.AnsiString, 255, navigationItem.URLHelp)
                                .AddParameter("SmallImage", DbType.AnsiString, 80, navigationItem.SmallImage)
                                .AddParameter("BigImage", DbType.AnsiString, 80, navigationItem.BigImage)
                                .AddParameter("Sequence", DbType.Decimal, 5, navigationItem.Sequence)
                                .AddParameter("RecordStatus", DbType.Decimal, 5, navigationItem.RecordStatus)
                                .AddParameter("UpdateUserCode", DbType.Decimal, 9, navigationItem.UpdateUserCode)
                                .AddParameter("UpdateDate", DbType.DateTime, 0, navigationItem.UpdateDate)
                                .AddParameter("NavigationId", DbType.Decimal, 9, navigationItem.NavigationId)
                                .Execute(connection, "Research");
        }

        /// <summary>
        /// Actualiza una lista de registros en la tabla Navigation por medio de su clave primaria.
        /// </summary>
        /// <remarks>Complemento para procesamiento masivo</remarks>
        /// <param name="navigationItems">Lista de instancia de Navigation</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista con el resultado de la creación de cada instancia.</returns>
        public static List<int> Update(List<Architect.API.Core.Contracts.General.Navigation> navigationItems, IDbConnection connection = null)
        {
            List<int> result = new List<int>();
            bool local = false;

            if (connection == null)
            {
                connection = Architect.DataFactory.Database.OpenConnection("Research");
                local = true;
            }
            foreach (Architect.API.Core.Contracts.General.Navigation item in navigationItems)
            {
                result.Add(Update(item, connection));
            }
            if (local)
            {
                connection.Close();
            }
            return result;
        }

        /// <summary>
        /// Elimina un registro en la tabla Navigation por medio de su clave primaria.
        /// </summary>
        /// <param name="navigationid">Identificación única del registro.</param>
        /// <param name="companyId">Identificación de la compañia propietaria.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros eliminados.</returns>
        public static int Delete(int navigationid, int companyId, IDbConnection connection = null)
        {
            return Database.Delete("DELETE FROM Navigation " +
                                    "WHERE NavigationId=:NavigationId AND CompanyId=:CompanyId")
                                .AddParameter("NavigationId", DbType.Decimal, 9, navigationid)
                                .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                                .Execute(connection, "Research");
        }

        /// <summary>
        /// Actualiza una lista de registros en la tabla Navigation por medio de su clave primaria.
        /// </summary>
        /// <remarks>Complemento para procesamiento masivo</remarks>
        /// <param name="navigationidList">Lista de Identificación única del registro.</param>
        /// <param name="companyId">Identificación de la compañia propietaria.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista con el resultado de la creación de cada instancia.</returns>
        public static List<int> Delete(List<int> navigationidList, int companyId, IDbConnection connection = null)
        {
            List<int> result = new List<int>();
            bool local = false;

            if (connection == null)
            {
                connection = Architect.DataFactory.Database.OpenConnection("Research");
                local = true;
            }
            foreach (int item in navigationidList)
            {
                result.Add(Delete(item, companyId, connection));
            }
            if (local)
            {
                connection.Close();
            }
            return result;
        }

        /// <summary>
        /// Recupera la cantidad de registros existentes en la tabla Navigation por medio de su clave primaria.
        /// </summary>
        /// <param name="navigationid">Identificación única del registro.</param>
        /// <param name="companyId">Identificación de la compañia propietaria.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros encontrados.</returns>
        public static int Count(int navigationid, int companyId, IDbConnection connection = null)
        {

            return (int)Database.Select("SELECT COUNT(NavigationId) " +
                                          "FROM Navigation " +
                                         "WHERE NavigationId=:NavigationId AND CompanyId=:CompanyId")
                               .AddParameter("NavigationId", DbType.Decimal, 9, navigationid)
                               .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                               .QueryScalar<Decimal>(connection, "Research");
        }

        /// <summary>
        /// Recupera la cantidad de registros existentes en la tabla Navigation que cumplen con el filtro.
        /// </summary>
        /// <param name="filter">Permite filtrar la lista de registros</param>
        /// <param name="parameters">Lista de parametros para complemento.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Cantidad de registros encontrados.</returns>
        public static int Count(string filter, List<DataFactory.Contracts.Parameter> parameters = null, IDbConnection connection = null)
        {
            return (int)Database.Select("SELECT COUNT(NavigationId) " +
                                          "FROM Navigation " + filter)
                               .AddParameter(parameters)
                               .QueryScalar<Decimal>(connection, "Research");
        }

        /// <summary>
        /// Recupera una lista para selección de la tabla Navigation.
        /// </summary>
        /// <param name="companyId">Identificación de la compañia propietaria.</param>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Lista de instancias de LookUpValue</returns>
        public static List<Utilities.Contracts.LookUpValue> LookUp(int companyId, IDbConnection connection = null)
        {
            List<Utilities.Contracts.LookUpValue> result = new List<Utilities.Contracts.LookUpValue>();
            Database.Select("SELECT NavigationId, Title " +
                              "FROM Navigation WHERE CompanyId=:CompanyId ORDER BY Title")
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                            {
                                result.Add(new Utilities.Contracts.LookUpValue()
                                {
                                    Code = reader.StringValue("NavigationId"),
                                    Description = reader.StringValue("Title")
                                });
                            }));
            return result;
        }

        /// <summary>
        /// Convierte una fila de un 'DataReader' a un clase de tipo 'Navigation'.
        /// </summary>
        /// <param name="reader">DataReader</param>
        /// <param name="item">Instancia pre creada</param>
        /// <returns>Instancia con la información de la fila del 'DataReader'</returns>
        public static Architect.API.Core.Contracts.General.Navigation DataReaderToNavigation(System.Data.IDataReader reader, Architect.API.Core.Contracts.General.Navigation item = null)
        {
            if (item == null)
            {
                item = new Architect.API.Core.Contracts.General.Navigation();
            }
            item.NavigationId = reader.IntegerValue("NavigationId");
            item.CompanyId = reader.IntegerValue("CompanyId");
            item.ParentCode = reader.StringValue("ParentCode");
            item.Code = reader.StringValue("Code");
            item.Type = reader.IntegerValue("Type");
            item.Title = reader.StringValue("Title");
            item.Description = reader.StringValue("Description");
            item.MenuTitle = reader.StringValue("MenuTitle");
            item.URLPath = reader.StringValue("URLPath");
            item.URLHelp = reader.StringValue("URLHelp");
            item.SmallImage = reader.StringValue("SmallImage");
            item.BigImage = reader.StringValue("BigImage");
            item.Sequence = reader.IntegerValue("Sequence");
            item.RecordStatus = reader.IntegerValue("RecordStatus");
            item.UpdateUserCode = reader.IntegerValue("UpdateUserCode");
            item.UpdateUserName = reader.StringValue("UpdateUserName");
            item.UpdateDate = reader.DateTimeValue("UpdateDate");
            return item;
        }

    }

}

File: Core/Architect.API.Core.DataAccess/_Generated/Navigation.vb
Imports Architect.Common.Proxy

Namespace Architect.API.Core.DataAccess.General

    ''' <summary>
    ''' 
    ''' </summary>
    Partial Public NotInheritable Class Navigation

        ''' <summary>
        ''' Crea un registro en la tabla Navigation.
        ''' </summary>
        ''' <param name="navigationItem">Instancia de Navigation</param>
        ''' <returns>Instancia de Navigation</returns>
        Public Shared Function Create(navigationItem As Architect.API.Core.Contracts.General.Navigation) As Integer
            Dim rows As Integer = 0
            With New DataManagerFactory("INSERT INTO Navigation " &
                                              "(NavigationId, Code, ParentCode, Type, Title, MenuTitle, Description, URLPath, URLHelp, SmallImage, BigImage, Sequence, RecordStatus, CompanyId, UpdateUserCode, UpdateDate) " &
                                        "VALUES(@:NavigationId, @:Code, @:ParentCode, @:Type, @:Title, @:MenuTitle, @:Description, @:URLPath, @:URLHelp, @:SmallImage, @:BigImage, @:Sequence, @:RecordStatus, @:CompanyId, @:UpdateUserCode, @:UpdateDate)", "Navigation", "Research")
                .AddParameter("NavigationId", DbType.Decimal, 9, navigationItem.NavigationId.IsEmpty(), navigationItem.NavigationId)
                .AddParameter("Code", DbType.AnsiStringFixedLength, 8, navigationItem.Code.IsEmpty(), navigationItem.Code)
                .AddParameter("ParentCode", DbType.AnsiStringFixedLength, 8, navigationItem.ParentCode.IsEmpty(), navigationItem.ParentCode)
                .AddParameter("Type", DbType.Decimal, 5, navigationItem.Type.IsEmpty(), navigationItem.Type)
                .AddParameter("Title", DbType.AnsiString, 120, navigationItem.Title.IsEmpty(), navigationItem.Title)
                .AddParameter("MenuTitle", DbType.AnsiString, 40, navigationItem.MenuTitle.IsEmpty(), navigationItem.MenuTitle)
                .AddParameter("Description", DbType.AnsiString, 255, navigationItem.Description.IsEmpty(), navigationItem.Description)
                .AddParameter("URLPath", DbType.AnsiString, 255, navigationItem.URLPath.IsEmpty(), navigationItem.URLPath)
                .AddParameter("URLHelp", DbType.AnsiString, 255, navigationItem.URLHelp.IsEmpty(), navigationItem.URLHelp)
                .AddParameter("SmallImage", DbType.AnsiString, 80, navigationItem.SmallImage.IsEmpty(), navigationItem.SmallImage)
                .AddParameter("BigImage", DbType.AnsiString, 80, navigationItem.BigImage.IsEmpty(), navigationItem.BigImage)
                .AddParameter("Sequence", DbType.Decimal, 5, navigationItem.Sequence.IsEmpty(), navigationItem.Sequence)
                .AddParameter("RecordStatus", DbType.Decimal, 5, navigationItem.RecordStatus.IsEmpty(), navigationItem.RecordStatus)
                .AddParameter("CompanyId", DbType.Decimal, 5, navigationItem.CompanyId.IsEmpty(), navigationItem.CompanyId)
                .AddParameter("UpdateUserCode", DbType.Decimal, 9, navigationItem.UpdateUserCode.IsEmpty(), navigationItem.UpdateUserCode)
                .AddParameter("UpdateDate", DbType.DateTime, 0, navigationItem.UpdateDate.IsEmpty(), navigationItem.UpdateDate)
                rows = .CommandExecute()
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Recupera un registro en la tabla Navigation por medio de su clave primaria.
        ''' </summary>
        ''' <param name="navigationid"></param>
        ''' <param name="companyId"></param>
        ''' <returns>Instancia de Navigation</returns>
        Public Shared Function Retrieve(navigationid As Integer, companyId As Integer) As Architect.API.Core.Contracts.General.Navigation
            Dim rows As Data.DataTable = Nothing
            Dim result As Architect.API.Core.Contracts.General.Navigation = Nothing
            With New DataManagerFactory("SELECT NavigationId, Code, ParentCode, Type, Title, MenuTitle, Description, URLPath, URLHelp, SmallImage, BigImage, Sequence, Navigation.RecordStatus, Navigation.CompanyId, Navigation.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Navigation.UpdateDate " &
                                          "FROM Navigation LEFT JOIN UserMember um ON um.UserId = Navigation.UpdateUserCode " &
                                         "WHERE NavigationId=@:NavigationId AND Navigation.CompanyId=@:CompanyId", "Navigation", "Research")
                .AddParameter("NavigationId", DbType.Decimal, 9, False, navigationid)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    result = DataRowMapper(rows.Rows(0))
                End If
            End With
            Return result
        End Function

        ''' <summary>
        ''' Recupera una lista de registros en la tabla Navigation.
        ''' </summary>
        ''' <param name="companyId"></param>
        ''' <param name="filter"></param>
        ''' <returns>Lista de instancias de Navigation</returns>
        Public Shared Function RetrieveAll(companyId As Integer, filter As String) As List(Of Architect.API.Core.Contracts.General.Navigation)
            Dim rows As Data.DataTable = Nothing
            Dim result As New List(Of Architect.API.Core.Contracts.General.Navigation)
            With New DataManagerFactory("SELECT NavigationId, Code, ParentCode, Type, Title, MenuTitle, Description, URLPath, URLHelp, SmallImage, BigImage, Sequence, Navigation.RecordStatus, Navigation.CompanyId, Navigation.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Navigation.UpdateDate " &
                                          "FROM Navigation LEFT JOIN UserMember um ON um.UserId = Navigation.UpdateUserCode " &
                                         "WHERE Navigation.CompanyId=@:CompanyId", "Navigation", "Research")
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    For Each itemRow As DataRow In rows.Rows
                        result.Add(DataRowMapper(itemRow))
                    Next
                End If
            End With
            Return result
        End Function

        ''' <summary>
        ''' Verifica si existe un registro en la tabla Navigation por medio de su clave primaria.
        ''' </summary>
        ''' <returns>Verdadero si existe el registro, en caso contrario falso.</returns>
        Public Shared Function RetrieveLastKey() As Integer
            Dim rows As Integer = 0

            With New DataManagerFactory("SELECT NVL(MAX(NavigationId),0) " &
                                          "FROM Navigation", "Navigation", "Research")
                rows = .QueryExecuteScalarToInteger
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Actualiza un registro en la tabla Navigation por medio de su clave primaria.
        ''' </summary>
        ''' <param name="navigationItem">Instancia de Navigation</param>
        ''' <param name="companyId"></param>
        ''' <returns>Cantidad de registros actualizados.</returns>
        Public Shared Function Update(navigationItem As Architect.API.Core.Contracts.General.Navigation, companyId As Integer) As Integer
            Dim rows As Integer = 0
            With New DataManagerFactory("UPDATE Navigation " &
                                           "SET Code=@:Code, ParentCode=@:ParentCode, Type=@:Type, Title=@:Title, MenuTitle=@:MenuTitle, Description=@:Description, URLPath=@:URLPath, URLHelp=@:URLHelp, SmallImage=@:SmallImage, BigImage=@:BigImage, Sequence=@:Sequence, RecordStatus=@:RecordStatus, CompanyId=@:CompanyId, UpdateUserCode=@:UpdateUserCode, UpdateDate=@:UpdateDate " &
                                         "WHERE NavigationId=@:NavigationId AND Navigation.CompanyId=@:CompanyId", "Navigation", "Research")
                .AddParameter("Code", DbType.AnsiStringFixedLength, 8, navigationItem.Code.IsEmpty(), navigationItem.Code)
                .AddParameter("ParentCode", DbType.AnsiStringFixedLength, 8, navigationItem.ParentCode.IsEmpty(), navigationItem.ParentCode)
                .AddParameter("Type", DbType.Decimal, 5, navigationItem.Type.IsEmpty(), navigationItem.Type)
                .AddParameter("Title", DbType.AnsiString, 120, navigationItem.Title.IsEmpty(), navigationItem.Title)
                .AddParameter("MenuTitle", DbType.AnsiString, 40, navigationItem.MenuTitle.IsEmpty(), navigationItem.MenuTitle)
                .AddParameter("Description", DbType.AnsiString, 255, navigationItem.Description.IsEmpty(), navigationItem.Description)
                .AddParameter("URLPath", DbType.AnsiString, 255, navigationItem.URLPath.IsEmpty(), navigationItem.URLPath)
                .AddParameter("URLHelp", DbType.AnsiString, 255, navigationItem.URLHelp.IsEmpty(), navigationItem.URLHelp)
                .AddParameter("SmallImage", DbType.AnsiString, 80, navigationItem.SmallImage.IsEmpty(), navigationItem.SmallImage)
                .AddParameter("BigImage", DbType.AnsiString, 80, navigationItem.BigImage.IsEmpty(), navigationItem.BigImage)
                .AddParameter("Sequence", DbType.Decimal, 5, navigationItem.Sequence.IsEmpty(), navigationItem.Sequence)
                .AddParameter("RecordStatus", DbType.Decimal, 5, navigationItem.RecordStatus.IsEmpty(), navigationItem.RecordStatus)
                .AddParameter("CompanyId", DbType.Decimal, 5, navigationItem.CompanyId.IsEmpty(), navigationItem.CompanyId)
                .AddParameter("UpdateUserCode", DbType.Decimal, 9, navigationItem.UpdateUserCode.IsEmpty(), navigationItem.UpdateUserCode)
                .AddParameter("UpdateDate", DbType.DateTime, 0, navigationItem.UpdateDate.IsEmpty(), navigationItem.UpdateDate)
                .AddParameter("NavigationId", DbType.Decimal, 9, False, navigationItem.NavigationId)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                rows = .CommandExecute()
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Elimina un registro en la tabla Navigation por medio de su clave primaria.
        ''' </summary>
        ''' <param name="navigationid"></param>
        ''' <param name="companyId"></param>
        ''' <returns>Cantidad de registros eliminados.</returns>
        Public Shared Function Delete(navigationid As Integer, companyId As Integer) As Integer
            Dim rows As Integer = 0
            With New DataManagerFactory("DELETE FROM Navigation " &
                                         "WHERE NavigationId=@:NavigationId AND CompanyId=@:CompanyId", "Navigation", "Research")
                .AddParameter("NavigationId", DbType.Decimal, 9, False, navigationid)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                rows = .CommandExecute()
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Recupera la cantidad de registros existentes en la tabla Navigation por medio de su clave primaria.
        ''' </summary>
        ''' <param name="navigationid"></param>
        ''' <returns>Cantidad de registros encontrados.</returns>
        Public Shared Function Count(navigationid As Integer) As Integer
            Dim rows As Integer = 0

            With New DataManagerFactory("SELECT COUNT(NavigationId) " &
                                          "FROM Navigation " &
                                         "WHERE NavigationId=@:NavigationId", "Navigation", "Research")
                .AddParameter("NavigationId", DbType.Decimal, 9, False, navigationid)
                rows = .QueryExecuteScalarToInteger
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Recupera una lista para selección de la tabla Navigation.
        ''' </summary>
        ''' <returns>Lista de instancias de LookUpValue</returns>
        Public Shared Function LookUp() As List(Of Architect.Common.DataType.LookUpValue)
            Dim rows As Data.DataTable = Nothing
            Dim result As New List(Of Architect.Common.DataType.LookUpValue)
            With New DataManagerFactory("SELECT NavigationId, Title " &
                                          "FROM Navigation ORDER BY Title", "Navigation", "Research")
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    For Each itemRow As DataRow In rows.Rows
                        result.Add(New Architect.Common.DataType.LookUpValue With {.Code = itemRow.StringValue("NavigationId"), .Description = itemRow.StringValue("Title")})
                    Next
                End If
            End With
            Return result
        End Function

#Region "Helper"

        ''' <summary>
        ''' Permite hacer el mapeo entre una fila de un datatable hacia una instancia de Navigation.
        ''' </summary>
        ''' <returns>Instancia de Navigation</returns>
        Private Shared Function DataRowMapper(itemRow As DataRow) As Architect.API.Core.Contracts.General.Navigation
            Return New Architect.API.Core.Contracts.General.Navigation With {.NavigationId = itemRow.NumericValue("NavigationId"), .Code = itemRow.StringValue("Code"), .ParentCode = itemRow.StringValue("ParentCode"), .Type = itemRow.NumericValue("Type"), .Title = itemRow.StringValue("Title"), .MenuTitle = itemRow.StringValue("MenuTitle"), .Description = itemRow.StringValue("Description"), .URLPath = itemRow.StringValue("URLPath"), .URLHelp = itemRow.StringValue("URLHelp"), .SmallImage = itemRow.StringValue("SmallImage"), .BigImage = itemRow.StringValue("BigImage"), .Sequence = itemRow.NumericValue("Sequence"), .RecordStatus = itemRow.NumericValue("RecordStatus"), .CompanyId = itemRow.NumericValue("CompanyId"), .UpdateUserCode = itemRow.NumericValue("UpdateUserCode"), .UpdateUserName = itemRow.StringValue("UpdateUserName"), .UpdateDate = itemRow.DateTimeValue("UpdateDate")}
        End Function

#End Region

    End Class

End Namespace

